Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,133 @@
+# DaemonSet deployment for Prometheus Agent
+
+* Owners:
+  * [haanhvu](https://github.com/haanhvu)
+* Related Tickets:
+  * [#5495](https://github.com/prometheus-operator/prometheus-operator/issues/5495)
+* Other docs:
+  * n/a
+
+This proposal is about designing and implementing the deployment of Prometheus Agent as DaemonSet. Currently, Prometheus Agent can only be deployed as StatefulSet, which could be considered as “cluster-wide” strategy, meaning one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster. Though this works well for many use cases, some use cases may indeed prefer a “node-specific” strategy (DaemonSet), where each Prometheus Agent only scrapes the metrics from the targets located on the same node.
+
+
+## 1. Why
+
+When deploying Prometheus Agent in Kubernetes, three of the biggest users’ concerns are: load distribution, scalability, and security.
+
+DaemonSet deployment solves all these three concerns:
+* Load distribution: Each Prometheus Agent pod will only scrape the targets located on the same node. Even though the targets on some nodes may produce more metrics than other nodes, the load distribution would be reliable enough.
+* Automatic scalability: When new nodes are added to the cluster, new Prometheus Agent pods will be automatically added in the nodes that meet user-defined restrictions (if any).
+* Security: Since the scraped targets are local to the Prometheus Agent pod (on the same node), the scope of security problems is reduced to each node.
+
+DaemonSet deployment is especially more suitable for Prometheus Agent than Prometheus, since the storage requirement for the Agent mode is pretty light compared to a fully functional Prometheus server.
+
+This deployment mode has been implemented in [Google Cloud Managed Service for Prometheus (GMP)](https://github.com/GoogleCloudPlatform/prometheus-engine/), so we have an implementation example to learn from. Also, since this deployment has been tested and proven with GMP’s userbase, we can count on that it can solve a large enough number of use cases.
```

Comment:
I have a feeling that we're going with a tone that may sound incorrect. It's true that Daemonset solves problems that are present with the Statefulset approach but at the cost of introducing new problems. It's not about Daemonset being better than Statefulset, but what problems we're trying to solve and what trade-offs users are willing to bear.

--- 

> Even though the targets on some nodes may produce more metrics than other nodes, the load distribution would be reliable enough.

* If the user's goal is to monitor only K8s master nodes, this is incorrect. Agents running on master nodes would be hot spots, while all others would remain idle. 
* If the user has metric producer applications that they can't afford to lose insight during cluster scale-ups/down, daemonsets could be a dangerous choice since storage is an unsolved problem.

> Since the scraped targets are local to the Prometheus Agent pod (on the same node), the scope of security problems is reduced to each node.

But it also introduces a new security flaw. Sensitive information is provided through PodMonitors to allow Prometheus to scrape pods in all nodes and those secrets are mounted to all Prometheus Agents, even though only a few might need them.

---

What do you think about this? Is daemonset set really a better option when compared to statefulset? Or is it just a game of trade-offs?

Maybe we could acknowledge that although Statefulset has been a good option for specific scenarios, there are other scenarios where daemonset shines, e.g. when the user cares more about simple scalability and not so much about durability.

Reply:
Yeah I never thought that one is better than the other in general. It's just better in some cases and some aspects. I'll try to make it more clear.